File: src/heimdall/snapshot.py
from __future__ import annotations

import hashlib
import shutil
from pathlib import Path

MAX_FILES = 2_000
MAX_FILE_BYTES = 8 * 1024 * 1024
MAX_TOTAL_BYTES = 32 * 1024 * 1024
IGNORED_PARTS = {".git", ".venv", ".pytest_cache", ".ruff_cache", "__pycache__"}


class SnapshotError(RuntimeError):
    def __init__(self, code: str):
        self.code = code
        super().__init__(code)


def _entries(root: Path):
    entries = []
    for path in root.rglob("*"):
        relative = path.relative_to(root)
        if any(part in IGNORED_PARTS for part in relative.parts):
            continue
        entries.append((relative, path))
    return sorted(entries, key=lambda item: item[0].as_posix().encode("utf-8"))


def tree_digest(root: str | Path) -> str:
    root_path = Path(root).resolve()
    if not root_path.is_dir():
        raise SnapshotError("target_unavailable")
    digest = hashlib.sha256()
    file_count = 0
    total_bytes = 0
    for relative, path in _entries(root_path):
        if path.is_symlink():
            raise SnapshotError("symlink_unsupported")
        relative_bytes = relative.as_posix().encode("utf-8")
        if path.is_dir():
            kind = b"D"
            payload = b""
        elif path.is_file():
            kind = b"F"
            file_count += 1
            if file_count > MAX_FILES:
                raise SnapshotError("target_too_large")
            size = path.stat().st_size
            if size > MAX_FILE_BYTES:
                raise SnapshotError("target_too_large")
            total_bytes += size
            if total_bytes > MAX_TOTAL_BYTES:
                raise SnapshotError("target_too_large")
            payload = path.read_bytes()
        else:
            raise SnapshotError("special_file_unsupported")
        digest.update(kind)
        digest.update(len(relative_bytes).to_bytes(8, "big"))
        digest.update(relative_bytes)
        digest.update(len(payload).to_bytes(8, "big"))
        digest.update(payload)
    return digest.hexdigest()


def copy_target(source: Path, destination: Path) -> None:
    tree_digest(source)
    shutil.copytree(
        source,
        destination,
        symlinks=False,
        ignore=shutil.ignore_patterns(*IGNORED_PARTS),
    )

File: internal/snapshot/snapshot.go
package snapshot

import (
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

const MaxFiles = 2000
const MaxFileBytes = 8 * 1024 * 1024
const MaxTotalBytes = 32 * 1024 * 1024

var ignored = map[string]bool{".git": true, ".venv": true, ".pytest_cache": true, ".ruff_cache": true, "__pycache__": true}

type Error struct{ code string }

func (err *Error) Error() string { return err.code }
func fail(code string) error     { return &Error{code: code} }
func Code(err error) string {
	var target *Error
	if errors.As(err, &target) {
		return target.code
	}
	return ""
}

type entry struct {
	relative, path string
	info           fs.FileInfo
}

func TreeDigest(root string) (string, error) {
	entries, err := collect(root)
	if err != nil {
		return "", err
	}
	digest := sha256.New()
	var fileCount, totalBytes int64
	var length [8]byte
	for _, item := range entries {
		kind := byte('D')
		payload := []byte{}
		if item.info.Mode().IsRegular() {
			kind = 'F'
			fileCount++
			if fileCount > MaxFiles || item.info.Size() > MaxFileBytes {
				return "", fail("target_too_large")
			}
			totalBytes += item.info.Size()
			if totalBytes > MaxTotalBytes {
				return "", fail("target_too_large")
			}
			payload, err = os.ReadFile(item.path)
			if err != nil {
				return "", fail("target_unavailable")
			}
		} else if !item.info.IsDir() {
			return "", fail("special_file_unsupported")
		}
		digest.Write([]byte{kind})
		binary.BigEndian.PutUint64(length[:], uint64(len([]byte(filepath.ToSlash(item.relative)))))
		digest.Write(length[:])
		digest.Write([]byte(filepath.ToSlash(item.relative)))
		binary.BigEndian.PutUint64(length[:], uint64(len(payload)))
		digest.Write(length[:])
		digest.Write(payload)
	}
	return hex.EncodeToString(digest.Sum(nil)), nil
}

func CopyTarget(source, destination string) error {
	if _, err := TreeDigest(source); err != nil {
		return err
	}
	rootInfo, err := os.Stat(source)
	if err != nil || !rootInfo.IsDir() {
		return fail("target_unavailable")
	}
	if err := os.Mkdir(destination, rootInfo.Mode().Perm()); err != nil {
		return err
	}
	entries, err := collect(source)
	if err != nil {
		return err
	}
	for _, item := range entries {
		target := filepath.Join(destination, item.relative)
		if item.info.IsDir() {
			if err := os.Mkdir(target, item.info.Mode().Perm()); err != nil {
				return err
			}
			continue
		}
		data, err := os.ReadFile(item.path)
		if err != nil {
			return err
		}
		if err := os.WriteFile(target, data, item.info.Mode().Perm()); err != nil {
			return err
		}
	}
	return nil
}

func collect(root string) ([]entry, error) {
	root, err := filepath.Abs(root)
	if err != nil {
		return nil, fail("target_unavailable")
	}
	info, err := os.Stat(root)
	if err != nil || !info.IsDir() {
		return nil, fail("target_unavailable")
	}
	var entries []entry
	err = filepath.WalkDir(root, func(path string, directory fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if path == root {
			return nil
		}
		relative, err := filepath.Rel(root, path)
		if err != nil {
			return err
		}
		for _, part := range strings.Split(filepath.ToSlash(relative), "/") {
			if ignored[part] {
				if directory.IsDir() {
					return filepath.SkipDir
				}
				return nil
			}
		}
		if directory.Type()&os.ModeSymlink != 0 {
			return fail("symlink_unsupported")
		}
		info, err := directory.Info()
		if err != nil {
			return err
		}
		entries = append(entries, entry{relative: relative, path: path, info: info})
		return nil
	})
	if err != nil {
		var snapshotError *Error
		if errors.As(err, &snapshotError) {
			return nil, err
		}
		return nil, fail("target_unavailable")
	}
	sort.Slice(entries, func(i, j int) bool {
		return filepath.ToSlash(entries[i].relative) < filepath.ToSlash(entries[j].relative)
	})
	return entries, nil
}
